File: code/python/剑指offer35.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""
class Solution:
    def copyRandomList(self, head: 'Node') -> 'Node':
        if head is None: return None
        originToCopy = {}
        temp = head
        last = None
        while temp is not None:
            node = Node(temp.val)
            if last is None:
                last = node
            else:
                last.next = node
                last = node
            originToCopy[temp] = node
            temp = temp.next

        temp = head
        while temp is not None:
            #print(temp.val,temp.random)
            if temp.random is None:
                originToCopy[temp].random = None
            else:
                originToCopy[temp].random = originToCopy[temp.random]
            temp = temp.next
        return originToCopy[head]

File: code/java/剑指offer35.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/
class Solution {
    public Node copyRandomList(Node head) {
        if(head==null) return null;
        Map<Node,Node> map = new HashMap<>();
        Node temp = head;
        Node last = null;
        while(temp!=null){
            Node node = new Node(temp.val);
            if(last==null) last = node;
            else{
                last.next = node;
                last = node;
            }
            map.put(temp,node);
            temp = temp.next;
        }
        temp = head;
        while(temp!=null){
            if(temp.random==null){
                map.get(temp).random=null;
            }else{
                map.get(temp).random = map.get(temp.random);
            }
            temp =temp.next;
        }
        return map.get(head);
    }
}